External/SPMmodified/xASL/xASL_im_FillNaNs.m

```diff
@@ -110,39 +111,45 @@
     IM(IM==0) = NaN;
 end
 
+% Also catch Infs or imaginary numbers or whatever. Zeroes are already catched above in the main function.
+IM(~isfinite(IM)) = NaN;
 
 %% ------------------------------------------------------------------------------
 %% 2. Replace NaNs using any of the methods
-fprintf('%s', 'Cleaning up NaNs in image:  0%');
-TotalDim = size(IM,4)*size(IM,5)*size(IM,6)*size(IM,7);
-CurrentIt = 1;
-for i4=1:size(IM,4)
-    for i5=1:size(IM,5)
-        for i6=1:size(IM,6)
-            for i7=1:size(IM,7)
-                xASL_TrackProgress(CurrentIt, TotalDim);
-
-                switch UseMethod
-                    case 1
-                        % Use this for cases where NaNs need to get border values (e.g. extrapolating a masked image)
-                        IM(:,:,:,i4,i5,i6,i7) = xASL_im_ExtrapolateSmoothOverNaNs(IM(:,:,:,i4,i5,i6,i7), bQuality, VoxelSize);
-                    case 2
-                        IM(isnan(IM)) = 0;
-                        % This loops unnecessary but goes fast enough, keep in this loop for readability
-                    case 3
-                        % Use this for cases where NaNs need to expand linearly (e.g., extrapolating flowfields)
-                        IM(:,:,:,i4,i5,i6,i7) = xASL_im_ExtrapolateLinearlyOverNaNs(IM(:,:,:,i4,i5,i6,i7));
-                    otherwise
-                        error('Invalid option');
-                end
-
-                CurrentIt = CurrentIt+1;
-            end
-        end
-    end
+% Run only if there were actual NaNs present
+if sum(isnan(IM(:))) > 0
+	switch UseMethod
+		case 1
+			% Method run is run independently for each 3D volume
+			fprintf('%s', 'Cleaning up NaNs in image:  0%');
+			TotalDim = size(IM,4)*size(IM,5)*size(IM,6)*size(IM,7);
+			CurrentIt = 1;
+			for i4=1:size(IM,4)
+				for i5=1:size(IM,5)
+					for i6=1:size(IM,6)
+						for i7=1:size(IM,7)
+							xASL_TrackProgress(CurrentIt, TotalDim);
+
+							% Use this for cases where NaNs need to get border values (e.g. extrapolating a masked image)
+							IM(:,:,:,i4,i5,i6,i7) = xASL_im_ExtrapolateSmoothOverNaNs(IM(:,:,:,i4,i5,i6,i7), bQuality, VoxelSize);
+							CurrentIt = CurrentIt+1;
+						end
+					end
+				end
+			end
+			fprintf('\n');
+		case 2
+			IM(isnan(IM)) = 0;
+		case 3
+			% Use this for cases where NaNs need to expand linearly (e.g., extrapolating deformation flowfields)
+			if ndims(IM) ~= 4 || size(IM, 4) ~= 3
```

Comment:
Incorrect. Deformation fields are always [X Y Z 1 3]

Reply:
See above - both options are possible. I have modified this and it now works for both.

Reply:
Duplicate to the other comment.